pensieve/statistics.py

```diff
@@ -0,0 +1,155 @@
+from abc import ABC, abstractmethod
+from decimal import Decimal
+import re
+from typing import Any, Dict, List, Optional
+
+import attr
+from google.cloud import bigquery
+import mozanalysis.bayesian_stats.bayesian_bootstrap as mabsbb
+from pandas import DataFrame
+
+
+@attr.s(auto_attribs=True)
+class StatisticResult:
+    """
+    Represents the resulting data after applying a statistic transformation
+    to metric data.
+    """
+
+    metric: str
+    statistic: str
+    parameter: Optional[Decimal]
+    label: str
+    ci_width: Optional[float] = 0.0
+    point: Optional[float] = 0.0
+    lower: Optional[float] = 0.0
+    upper: Optional[float] = 0.0
+
+
+@attr.s(auto_attribs=True)
+class StatisticResultCollection:
+    """
+    Represents a set of statistics result data.
+    """
+
+    data: List[StatisticResult] = []
+
+    def save_to_bigquery(self, client, destination_table, append=True):
+        """Stores the data to a BigQuery table with a defined schema."""
+
+        job_config = bigquery.LoadJobConfig()
+        job_config.schema = [
+            bigquery.SchemaField("metric", "STRING"),
+            bigquery.SchemaField("statistic", "STRING"),
+            bigquery.SchemaField("parameter", "NUMERIC"),
+            bigquery.SchemaField("label", "STRING"),
+            bigquery.SchemaField("ci_width", "FLOAT64"),
+            bigquery.SchemaField("point", "FLOAT64"),
+            bigquery.SchemaField("lower", "FLOAT64"),
+            bigquery.SchemaField("upper", "FLOAT64"),
+        ]
+
+        if append:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_APPEND
+        else:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        client.load_table_from_json(
+            self.to_dict()["data"], destination_table, job_config=job_config
+        )
+
+    def to_dict(self):
+        """Return statistic results as dict."""
+
+        return attr.asdict(self)
+
+
+@attr.s(auto_attribs=True)
+class Statistic(ABC):
+    """
+    Abstract representation of a statistic.
+
+    A statistic is a transformation that accepts a table of per-client aggregates and
+    returns a table representing a summary of the aggregates with respect to the branches
+    of the experiment.
+    """
+
+    ref_branch_label: str = "control"
```

Comment:
Is it fair to assume that every statistic has `ref_branch_label`?

Reply:
Not every, but many statistics will be interested in making a comparison to the control group.